File: gerenciamento.js
// Classe Contato que representa um contato com nome, telefone e email.
class Contato {
    constructor(nome, telefone, email) {
        this.nome = nome;
        this.telefone = telefone;
        this.email = email;
    }

    toString() {
        return `Nome: ${this.nome}, Telefone: ${this.telefone}, Email: ${this.email}`;
    }
}

// Classe GerenciadorContatos que gerencia a lista de contatos.
class GerenciadorContatos {
    constructor() {
        this.contatos = [];
    }

    adicionarContato(contato) {
        this.contatos.push(contato);
    }

    removerContato(nome) {
        this.contatos = this.contatos.filter(contato => contato.nome !== nome);
    }

    listarContatos() {
        this.contatos.forEach(contato => console.log(contato.toString()));
    }

    // Aplicando o padrão Strategy para buscar contatos
    setEstrategiaBusca(estrategia) {
        this.estrategiaBusca = estrategia;
    }

    buscarContato(nome) {
        return this.estrategiaBusca.buscar(this.contatos, nome);
    }
}

// Padrão Decorator aplicado para estender o GerenciadorContatos
class GerenciadorContatosDecorator {
    constructor(gerenciador) {
        this.gerenciador = gerenciador;
    }

    adicionarContato(contato) {
        this.gerenciador.adicionarContato(contato);
    }

    removerContato(nome) {
        this.gerenciador.removerContato(nome);
    }

    listarContatos() {
        this.gerenciador.listarContatos();
    }

    buscarContato(nome) {
        return this.gerenciador.buscarContato(nome);
    }
}

// Estratégia de busca por nome
class EstrategiaBuscaPorNome {
    buscar(contatos, nome) {
        return contatos.filter(contato => contato.nome.toLowerCase() === nome.toLowerCase());
    }
}

// Interface de linha de comando (CLI)
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const gerenciador = new GerenciadorContatos();
gerenciador.setEstrategiaBusca(new EstrategiaBuscaPorNome());

function mostrarMenu() {
    console.log(`
        Escolha uma opção:
        1. Adicionar contato
        2. Remover contato
        3. Listar contatos
        4. Buscar contato por nome
        5. Sair
    `);
}

function adicionarContato() {
    rl.question('Nome: ', nome => {
        rl.question('Telefone: ', telefone => {
            rl.question('Email: ', email => {
                const contato = new Contato(nome, telefone, email);
                gerenciador.adicionarContato(contato);
                console.log('Contato adicionado com sucesso!');
                iniciar();
            });
        });
    });
}

function removerContato() {
    rl.question('Nome do contato a ser removido: ', nome => {
        gerenciador.removerContato(nome);
        console.log('Contato removido com sucesso!');
        iniciar();
    });
}

function listarContatos() {
    console.log('Lista de Contatos:');
    gerenciador.listarContatos();
    iniciar();
}

function buscarContato() {
    rl.question('Nome do contato a buscar: ', nome => {
        const resultado = gerenciador.buscarContato(nome);
        if (resultado.length > 0) {
            console.log('Contatos encontrados:');
            resultado.forEach(contato => console.log(contato.toString()));
        } else {
            console.log('Contato não encontrado.');
        }
        iniciar();
    });
}

function iniciar() {
    mostrarMenu();
    rl.question('Opção: ', opcao => {
        switch(opcao) {
            case '1':
                adicionarContato();
                break;
            case '2':
                removerContato();
                break;
            case '3':
                listarContatos();
                break;
            case '4':
                buscarContato();
                break;
            case '5':
                rl.close();
                break;
            default:
                console.log('Opção inválida!');
                iniciar();
        }
    });
}

iniciar();

File: gerenciamento.py
# Classe Contato
class Contato:
    def __init__(self, nome, telefone, email):
        self.nome = nome
        self.telefone = telefone
        self.email = email

    def __str__(self):
        return f"Nome: {self.nome}, Telefone: {self.telefone}, Email: {self.email}"


# Padrão Strategy para busca de contatos
from abc import ABC, abstractmethod


class EstrategiaBusca(ABC):
    @abstractmethod
    def buscar(self, contatos, busca):
        pass


class EstrategiaBuscaPorNome(EstrategiaBusca):
    def buscar(self, contatos, busca):
        return [
            contato for contato in contatos if busca.lower() in contato.nome.lower()
        ]


# Classe GerenciadorContatos
class GerenciadorContatos:
    def __init__(self):
        self.contatos = []
        self.estrategia_busca = None

    def adicionar_contato(self, contato):
        self.contatos.append(contato)

    def remover_contato(self, nome):
        self.contatos = [contato for contato in self.contatos if contato.nome != nome]

    def listar_contatos(self):
        for contato in self.contatos:
            print(contato)

    def set_estrategia_busca(self, estrategia):
        self.estrategia_busca = estrategia

    def buscar_contato(self, busca):
        if not self.estrategia_busca:
            print("Estratégia de busca não definida.")
            return []
        return self.estrategia_busca.buscar(self.contatos, busca)


# Padrão Decorator
class DecoratorGerenciadorContatos:
    def __init__(self, gerenciador):
        self.gerenciador = gerenciador

    def adicionar_contato(self, contato):
        self.gerenciador.adicionar_contato(contato)

    def remover_contato(self, nome):
        self.gerenciador.remover_contato(nome)

    def listar_contatos(self):
        self.gerenciador.listar_contatos()

    def set_estrategia_busca(self, estrategia):
        self.gerenciador.set_estrategia_busca(estrategia)

    def buscar_contato(self, busca):
        return self.gerenciador.buscar_contato(busca)


# Decorator  que adiciona funcionalidade de logging
class GerenciadorComLog(DecoratorGerenciadorContatos):
    def adicionar_contato(self, contato):
        print(f"[LOG] Adicionando contato: {contato.nome}")
        super().adicionar_contato(contato)

    def remover_contato(self, nome):
        print(f"[LOG] Removendo contato: {nome}")
        super().remover_contato(nome)


# Interface do menu
def mostrar_menu():
    print(
        """
Escolha uma opção:
1. Adicionar contato
2. Remover contato
3. Listar contatos
4. Buscar contato por nome
5. Sair
"""
    )


def adicionar_contato(gerenciador):
    nome = input("Nome: ")
    telefone = input("Telefone: ")
    email = input("Email: ")
    contato = Contato(nome, telefone, email)
    gerenciador.adicionar_contato(contato)
    print("Contato adicionado com sucesso")


def remover_contato(gerenciador):
    nome = input("Nome do contato a ser removido: ")
    gerenciador.remover_contato(nome)
    print("Contato removido com sucesso")


def listar_contatos(gerenciador):
    print("Lista de Contatos:")
    gerenciador.listar_contatos()


def buscar_contato(gerenciador):
    busca = input("Nome do contato a buscar: ")
    resultados = gerenciador.buscar_contato(busca)
    if resultados:
        print("Contatos encontrados:")
        for contato in resultados:
            print(contato)
    else:
        print("Contato não encontrado.")


def main():
    gerenciador = GerenciadorContatos()
    gerenciador_com_log = GerenciadorComLog(gerenciador)
    estrategia_busca = EstrategiaBuscaPorNome()
    gerenciador_com_log.set_estrategia_busca(estrategia_busca)

    while True:
        mostrar_menu()
        opcao = input("Opção: ")

        if opcao == "1":
            adicionar_contato(gerenciador_com_log)
        elif opcao == "2":
            remover_contato(gerenciador_com_log)
        elif opcao == "3":
            listar_contatos(gerenciador_com_log)
        elif opcao == "4":
            buscar_contato(gerenciador_com_log)
        elif opcao == "5":
            print("Saindo do sistema.")
            break
        else:
            print("Opção inválida. Por favor, tente novamente.")


if __name__ == "__main__":
    main()
